File: 백준/Gold/7576. 토마토/토마토.py
from sys import stdin
from collections import deque

# 입력 처리
M, N = map(int, stdin.readline().split())
matrix = [list(map(int, stdin.readline().split())) for _ in range(N)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def bfs(dq):
    while dq:
        x, y = dq.popleft()
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            # 격자 범위를 벗어나지 않으면서, 익지 않은 토마토인 경우
            if 0 <= nx < N and 0 <= ny < M and matrix[nx][ny] == 0:
                matrix[nx][ny] = matrix[x][y] + 1
                dq.append((nx, ny))

def get_ripe_date():
    days = 0
    for line in matrix:
        for tomato in line:
            if tomato == 0:
                return 0
        days = max(days, max(line))
    return days

dq = deque()

for i in range(N):
    for j in range(M):
        if matrix[i][j] == 1:
            dq.append((i, j))

bfs(dq)

print(get_ripe_date() - 1)

File: 백준/Gold/7576. 토마토/토마토.java
import java.io.*;
import java.util.*;

public class Main {
    static int M, N;
    static int[][] matrix;
    static int[] dx = { -1, 1, 0, 0 };
    static int[] dy = { 0, 0, -1, 1 };

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        M = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());

        matrix = new int[N][M];

        Queue<int[]> queue = new LinkedList<>();

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                matrix[i][j] = Integer.parseInt(st.nextToken());
                if (matrix[i][j] == 1) {
                    queue.offer(new int[] { i, j });
                }
            }
        }

        bfs(queue);

        int days = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (matrix[i][j] == 0) {
                    System.out.println(-1);
                    return;
                }
                days = Math.max(days, matrix[i][j]);
            }
        }

        System.out.println(days - 1);
    }

    static void bfs(Queue<int[]> queue) {
        while (!queue.isEmpty()) {
            int[] pos = queue.poll();
            int x = pos[0], y = pos[1];

            for (int i = 0; i < 4; i++) {
                int nx = dx[i] + x, ny = dy[i] + y;

                if (nx >= 0 && nx < N && ny >= 0 && ny < M && matrix[nx][ny] == 0) {
                    matrix[nx][ny] = matrix[x][y] + 1;
                    queue.offer(new int[] { nx, ny });
                }
            }
        }
    }
}